File: Binary Tree/c++/depth_first_values.cpp
#include<bits/stdc++.h>
using namespace std;

class Node
{
    public:
        string val;
        Node* left;
        Node* right;

        Node(string initialVal)
        {
            val=initialVal;
            left=nullptr;
            right=nullptr;
        }
};

std::vector<std::string> depthFirstValues(Node* root)
{
    if (root)
    {
        vector<string>re{root->val};
        vector<string>left=depthFirstValues(root->left);
        vector<string>right=depthFirstValues(root->right);
        re.insert(re.end(),left.begin(),left.end());
        re.insert(re.end(),right.begin(),right.end());

        return re;
    }

    return {};
}

std::vector<std::string> depthFirstValues_Stack(Node* root)
{
    if (root)
    {
        vector<string>re;
        stack<Node*>s;
        s.push(root);

        while(s.size())
        {
            Node* current=s.top();
            s.pop();
            re.push_back(current->val);
            if (current->right) s.push(current->right);
            if (current->left) s.push(current->left);
        }

        return re;
    }

    return {};
}

int main()
{

}

//https://structy.net/problems/depth-first-values

File: Binary Tree/python/depth_first_values.py
class Node:
    def __init__(self, val):
        self.val=val
        self.left=None
        self.right=None

def depth_first_values(root):#nlr
    # print(root.val,root.left,root.right)
    if root:
        # re=[root.val]
        # re+=depth_first_values(root.left)
        # re+=depth_first_values(root.right)
        # return re
        return [root.val,*depth_first_values(root.left),*depth_first_values(root.right)]

    return []

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
g = Node('g')
a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g

print(depth_first_values(a))